### PR TITLE
PARQUET-501: Add OutputStream abstract interface, refactor encoding code paths

### PR DESCRIPTION
**Pull request by wesm**
I also did a bit of tidying / reorganization and giving interfaces more descriptive names.

### CODE REVIEW

**Comment by majetideepak** [comment]
This will be some conflict with #30.

File: src/parquet/util/test-common.h [hunk 1]
```diff
@@ -29,6 +29,16 @@ namespace parquet_cpp {
 namespace test {
 
 template <typename T>
+static inline void assert_vector_equal(const vector<T>& left,
+    const vector<T>& right) {
+  ASSERT_EQ(left.size(), right.size());
+
+  for (size_t i = 0; i < left.size(); ++i) {
+    ASSERT_EQ(left[i], right[i]) << i;
```
Comment (majetideepak) [inline src/parquet/util/test-common.h]: What does this shift do?

File: src/parquet/util/input.h [hunk 2]
```diff
@@ -15,15 +15,63 @@
...
+  void CloseFile();
+
+  std::string path_;
+  FILE* file_;
+  bool is_open_;
+};
+
+// ----------------------------------------------------------------------
+// Streaming input interfaces
+
```
Comment (majetideepak) [inline src/parquet/util/input.h]: Can you modify the `Peek` and `Read` APIs in `InputStream` and InMemoryInputStreamwith int64_t ?

File: src/parquet/util/test-common.h (hunk 1 above)
Comment (wesm) [inline src/parquet/util/test-common.h]: It prints `i` on assertion failure

File: src/parquet/util/input.h (hunk 2 above)
Comment (wesm) [inline src/parquet/util/input.h]: Outside the scope of this patch, but alright

Comment (wesm) [inline src/parquet/util/input.h]: done

**Comment by wesm** [comment]
I squashed these commits to simplify cherry-picking because #47 depends on this patch.

**Comment by wesm** [comment]
Rebased. @asandryh and @majetideepak let me know any other comments

**Comment by julienledem** [comment]
+1 LGTM

**Pull request closed without merging**